buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/java/AvailableJavaInstallations.kt

```diff
@@ -66,71 +63,83 @@ private
+    fun validateForAllBuilds() {
+        validate(validateBuildJdks())
     }
 
-    fun validateForRemoteBuildCacheUsage() {
-        val validationErrors = validateCompilationJdks()
-        if (validationErrors.isNotEmpty()) {
-            val message = formatValidationError(
-                "In order to have cache hits from the remote build cache, your environment needs to be configured accordingly!",
-                validationErrors
-            )
-            throw GradleException(message)
+    fun validateForCompilation() {
+        if (remoteBuildCacheEnabled()) {
+            validate(validateForRemoteCache())
         }
+        validate(validateCompilationJdks())
     }
 
     fun validateForProductionEnvironment() {
-        val validationErrors = validateCompilationJdks() +
-            mapOf(
-                validationMessage(testJavaHomePropertyName, javaInstallationForTest, oracleJdk8) to (javaInstallationForTest.vendorAndMajorVersion != oracleJdk8)
-            ).filterValues { it }.keys
-        if (validationErrors.isNotEmpty()) {
-            throw GradleException(formatValidationError("JDKs not configured correctly for production build.", validationErrors))
+        validate(validateProductionJdks())
+    }
+
+    private
+    fun remoteBuildCacheEnabled() = (project.gradle as GradleInternal).settings.buildCache.remote?.isEnabled == true
+
+    private
+    fun validateForRemoteCache(): Map<String, Boolean> =
+        mapOf("Must use Oracle JDK 8/9 to perform this build. Is currently ${currentJavaInstallation.vendorAndMajorVersion} at ${currentJavaInstallation.javaHome}." to
+            (currentJavaInstallation.vendorAndMajorVersion != oracleJdk8 && currentJavaInstallation.vendorAndMajorVersion != oracleJdk9))
+
+    private
+    fun validate(errorMessages: Map<String, Boolean>) {
+        val errors = errorMessages.filterValues { it }.keys
+        if (errors.isNotEmpty()) {
+            throw GradleException(formatValidationError("JDKs not configured correctly for the build.", errors))
         }
     }
 
     private
-    fun validateCompilationJdks(): Collection<String> {
-        val jdkForCompilation = javaInstallations.values.firstOrNull()
-        return mapOf(
-            "Must set project or system property '$java7HomePropertyName' to the path of an $oracleJdk7, is currently unset." to (jdkForCompilation == null),
-            validationMessage(java7HomePropertyName, jdkForCompilation, oracleJdk7) to (jdkForCompilation != null && jdkForCompilation.vendorAndMajorVersion != oracleJdk7),
+    fun validateCompilationJdks(): Map<String, Boolean> =
+        mapOf(
+            "Must use JDK 9+ to perform compilation in this build. It's currently ${javaInstallationForCompilation.vendorAndMajorVersion} at ${javaInstallationForCompilation.javaHome}. " +
+                "You can set a project or system property '$java9HomePropertyName' to an Java9-compatible JDK home path" to
```

Comment:
I guess the message should also say that it is possible to run the build on a Java 9 compatible JDK.

Reply:
Done.